src/Dofs/sparsity_pattern.jl

```diff
@@ -102,38 +415,164 @@ function _coupling_to_local_dof_coupling(dh::DofHandler, coupling::AbstractMatri
     return outs
 end
 
-"""
-    _add_cross_coupling(coupling_sdh, dof_i, dof_j, cell_field_dofs, neighbor_field_dofs, i, j, sym, keep_constrained, ch, cnt, I, J)
+function _create_sparsity_pattern!(
+        sp::AbstractSparsityPattern, dh::DofHandler, ch::Union{ConstraintHandler, Nothing},
+        keep_constrained::Bool,
+        coupling::Union{Vector{<:AbstractMatrix{Bool}}, Nothing},
+        cross_coupling::Union{AbstractMatrix{Bool}, Nothing},
+        topology,
+    )
+    # 1. Add all connections between dofs for every cell while filtering based
+    #    on a) constraints, and b) field/dof coupling.
+    cc = CellCache(dh)
+    for (sdhi, sdh) in pairs(dh.subdofhandlers)
+        set = BitSet(sdh.cellset)
+        coupling === nothing || (coupling_sdh = coupling[sdhi])
+        for cell_id in set
+            reinit!(cc, cell_id)
+            for (i, row) in pairs(cc.dofs)
+                # a) check constraint for row
+                !keep_constrained && haskey(ch.dofmapping, row) && continue
+                # TODO: Extracting the row here and reinserting after the j-loop
+                #       should give some nice speedup
+                for (j, col) in pairs(cc.dofs)
+                    # b) check coupling between (local) dofs i and j
+                    coupling === nothing || coupling_sdh[i, j] || continue
+                    # a) check constraint for col
+                    !keep_constrained && haskey(ch.dofmapping, col) && continue
+                    # Insert col as a non zero index for this row
+                    add_entry!(sp, row, col)
+                end
+            end
+        end
+    end
+    # 2. Make sure diagonal entries are included
+    for d in 1:ndofs(dh)
+        add_entry!(sp, d, d)
+    end
```

Comment:
Why? For example many saddle point problems do not need these for the block 2,2.

Reply:
Many sparse solvers need them.

Reply:
Sparse solvers as in direct sparse solvers?

Reply:
This isn't new, we already added the diagonal explicitly. `apply!` actually requires it too, since we insert a dummy equation on constrained dofs. Even if that wasn't the case it doesn't really cost anything to have it.